src/prefect/engine/flow_runner.py

```diff
@@ -438,21 +435,42 @@ def extra_context(task: Task, task_index: int = None) -> dict:
 
                 # this dictionary is used exclusively for "reduce" tasks in particular we store
                 # the states / futures corresponding to the upstream children, and if running
-                # on Dask, let Dask resolve them at the appropriate time
+                # on Dask, let Dask resolve them at the appropriate time.
+                # Note: this is an optimization that allows Dask to resolve the mapped
+                # dependencies by "elevating" them to a function argument.
                 upstream_mapped_states = {}  # type: Dict[Edge, list]
 
                 # -- process each edge to the task
                 for edge in self.flow.edges_to(task):
+
+                    # load the upstream task states (supplying Pending as a default)
                     upstream_states[edge] = task_states.get(
                         edge.upstream_task, Pending(message="Task state not available.")
                     )
 
+                    # if the edge is flattened and not the result of a map, then we
+                    # preprocess the upstream states. If it IS the result of a
+                    # map, it will be handled in `prepare_upstream_states_for_mapping`
+                    if edge.flattened:
+                        if not isinstance(upstream_states[edge], Mapped):
+                            upstream_states[edge] = executor.submit(
+                                executors.flatten_upstream_state, upstream_states[edge]
+                            )
+
                     # this checks whether the task is a "reduce" task for a mapped pipeline
                     # and if so, collects the appropriate upstream children
                     if not edge.mapped and isinstance(upstream_states[edge], Mapped):
-                        upstream_mapped_states[edge] = mapped_children.get(
-                            edge.upstream_task, []
-                        )
+                        children = mapped_children.get(edge.upstream_task, [])
+
+                        # if the edge is flattened, then we need to wait for the mapped children
+                        # to complete and then flatten them
+                        if edge.flattened:
+                            children = executor.wait(children)
```

Comment:
IIUC, this will pull retrieve the mapped results locally, defeating the point of computing the lengths remotely. I *think* you can safely remove this `wait` call and everything will still work.

Reply:
Thanks for spotting that! I think this got left behind from an earlier attempt, and I've removed it.